File: harness/ts/agent/agent.ts
// Agent base class. Every TS agent under agents/<id>/agent.ts exports a
// default class that extends Agent and implements run().
//
// The harness invokes:
//
//   const agent = new MyAgent();
//   const traj = await agent.run(goal, browserSession, budget, ctx);
//
// `ctx` carries the per-task identifiers needed to construct the trajectory
// file path runs/<agent>/<task>/<seed>/trajectory.jsonl[.gz]. The first three
// args (goal, browser, budget) are the contract spelled out in US-002.

import type { BrowserSession, Budget } from "./types.js";
import type { Trajectory } from "./trajectory.js";

export interface AgentContext {
  /** Stable identifier for the task being run. */
  task_id: string;
  /** Seed (an integer); rerunning with the same seed should reproduce. */
  seed: number;
  /** Root for trajectory output, typically <repo>/runs. */
  runs_root: string;
}

export abstract class Agent {
  /** Stable identifier matching agents/<id>/manifest.yaml. */
  abstract readonly id: string;

  abstract run(
    goal: string,
    browser: BrowserSession,
    budget: Budget,
    ctx: AgentContext,
  ): Promise<Trajectory>;
}

File: harness/python/gba_agent/agent.py
"""Python Agent base class. Subclass and implement ``run``.

The harness instantiates the agent, then invokes::

    agent.run(goal, browser, budget, trajectory, ctx)

``browser`` / ``budget`` / ``trajectory`` are RPC proxies that route calls
back to the TS harness; ``ctx`` carries per-task identifiers needed to
correlate the trajectory file path.
"""

from __future__ import annotations

from dataclasses import dataclass

from .browser import BrowserProxy
from .budget import Budget
from .trajectory import TrajectoryProxy


@dataclass(frozen=True)
class AgentContext:
    agent_id: str
    task_id: str
    seed: int


class Agent:
    """Base class. Concrete agents must set ``id`` and override ``run``."""

    id: str = "unnamed"

    def run(
        self,
        goal: str,
        browser: BrowserProxy,
        budget: Budget,
        trajectory: TrajectoryProxy,
        ctx: AgentContext,
    ) -> None:
        raise NotImplementedError
